# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/dns/DNSPacket.cpp
# Commit: 6f4ca967aa44
# Full Hash: 6f4ca967aa446cf24f4224576f51803c982f3d7b
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2024-06-14 20:49:45
# Regressor Bug: 1901990
# File Overlap Count: 2
# Description:
#   Bug 1901990 - Introduce a pref for setting a mock HTTPS RR, r=necko-reviewers,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D213559
# ==============================================================================

diff -r 8a2bdb44382a -r 6f4ca967aa44 netwerk/dns/DNSPacket.cpp
--- a/netwerk/dns/DNSPacket.cpp	Fri Jun 14 13:27:36 2024 +0000
+++ b/netwerk/dns/DNSPacket.cpp	Fri Jun 14 13:42:37 2024 +0000
@@ -327,33 +327,12 @@
 
 const uint8_t kDNS_CLASS_IN = 1;
 
-nsresult DNSPacket::EncodeRequest(nsCString& aBody, const nsACString& aHost,
-                                  uint16_t aType, bool aDisableECS) {
-  aBody.Truncate();
-  // Header
-  aBody += '\0';
-  aBody += '\0';  // 16 bit id
-  aBody += 0x01;  // |QR|   Opcode  |AA|TC|RD| Set the RD bit
-  aBody += '\0';  // |RA|   Z    |   RCODE   |
-  aBody += '\0';
-  aBody += 1;  // QDCOUNT (number of entries in the question section)
-  aBody += '\0';
-  aBody += '\0';  // ANCOUNT
-  aBody += '\0';
-  aBody += '\0';  // NSCOUNT
-
-  char additionalRecords =
-      (aDisableECS || StaticPrefs::network_trr_padding()) ? 1 : 0;
-  aBody += '\0';               // ARCOUNT
-  aBody += additionalRecords;  // ARCOUNT low byte for EDNS(0)
-
-  // Question
-
+// static
+nsresult DNSPacket::EncodeHost(nsCString& aBody, const nsACString& aHost) {
   // The input host name should be converted to a sequence of labels, where
   // each label consists of a length octet followed by that number of
   // octets.  The domain name terminates with the zero length octet for the
   // null label of the root.
-  // Followed by 16 bit QTYPE and 16 bit QCLASS
 
   int32_t index = 0;
   int32_t offset = 0;
@@ -383,6 +362,37 @@
     offset += labelLength + 1;  // move over label and dot
   } while (true);
 
+  return NS_OK;
+}
+
+nsresult DNSPacket::EncodeRequest(nsCString& aBody, const nsACString& aHost,
+                                  uint16_t aType, bool aDisableECS) {
+  aBody.Truncate();
+  // Header
+  aBody += '\0';
+  aBody += '\0';  // 16 bit id
+  aBody += 0x01;  // |QR|   Opcode  |AA|TC|RD| Set the RD bit
+  aBody += '\0';  // |RA|   Z    |   RCODE   |
+  aBody += '\0';
+  aBody += 1;  // QDCOUNT (number of entries in the question section)
+  aBody += '\0';
+  aBody += '\0';  // ANCOUNT
+  aBody += '\0';
+  aBody += '\0';  // NSCOUNT
+
+  char additionalRecords =
+      (aDisableECS || StaticPrefs::network_trr_padding()) ? 1 : 0;
+  aBody += '\0';               // ARCOUNT
+  aBody += additionalRecords;  // ARCOUNT low byte for EDNS(0)
+
+  // Question
+
+  nsresult rv = EncodeHost(aBody, aHost);
+  if (NS_FAILED(rv)) {
+    return rv;
+  }
+
+  // Followed by 16 bit QTYPE and 16 bit QCLASS
   aBody += static_cast<uint8_t>(aType >> 8);  // upper 8 bit TYPE
   aBody += static_cast<uint8_t>(aType);
   aBody += '\0';           // upper 8 bit CLASS